File: 실버 1 (7문제)/1697_숨바꼭질.py
import collections

# declare : queue
queue = collections.deque()

# N : my position
# K : brother's position
N, K = map(int, input().split())

# brother is to the left of me
if N >= K:
    print(N - K)
# brother is to the right of me
else:
    # declare : max number can be used in this algorithm
    MAX = 150000

    # declare : list that contains distance from N
    distance = [0 for i in range(MAX)]

    # initialize : queue
    queue.append([N, 0])

    # iterates the loop until queue is not empty
    while queue:
        # get : current position, current count from N
        cur_position, cur_count = queue.popleft()

        if cur_position == K:
            distance[cur_position] = cur_count
            break

        # continue : cur_position is out of valid coordinates
        if cur_position < 0 or cur_position >= MAX:
            continue

        # if pos(cur_position - 1) is not visited
        if cur_position - 1 >= 0 and distance[cur_position - 1] == 0:
            # update pos(cur_position - 1)
            distance[cur_position - 1] = cur_count + 1

            # add : add list of [cur_position - 1, cur_count + 1] to queue
            queue.append([cur_position - 1, cur_count + 1])

        # if pos(cur_position + 1) is not visited
        if cur_position + 1 < MAX and distance[cur_position + 1] == 0:
            # update pos(cur_position + 1)
            distance[cur_position + 1] = cur_count + 1

            # add : add list of [cur_position + 1, cur_count + 1] to queue
            queue.append([cur_position + 1, cur_count + 1])

        # if pos(cur_position * 2) is not visited
        if cur_position * 2 < MAX and distance[cur_position * 2] == 0:
            # update pos(cur_position * 2)
            distance[cur_position * 2] = cur_count + 1

            # add : add list of [cur_position * 2, cur_count + 1] to queue
            queue.append([cur_position * 2, cur_count + 1])

    print(distance[K])

File: 실버 1 (7문제)/1697_숨바꼭질.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 150000

// declare : structure for queue
typedef struct node {
	int cur_pos;
	int count;
}node;

// declare : queue, size of MAX
node queue[MAX];

// declare : queue variables
int front = -1, rear = -1;

// declare : queue functions
void enque(int cur_pos, int count);
node deque();

// declare : array recording distances
int distance[MAX];

// declare : N - my position, K - brother position
int N, K;

// declare : function that returns min(a, b)
int get_min(int a, int b);

int main() {
	// get input : N, K
	scanf("%d %d", &N, &K);

	// brother is to the left of me
	if (N >= K) {
		printf("%d", N - K);
	}
	// brother is to the right of me
	else {
		// initialize : queue
		enque(N, 0);

		// bfs
		while (front != rear) {
			// get : current node
			node temp = deque();

			// get : current pos, count
			int new_pos = temp.cur_pos;
			int new_count = temp.count;

			// continue : out of validate coordinates
			if (new_pos >= MAX || new_pos < 0) {
				continue;
			}

			if (new_pos == K) {
				break;
			}
				
			// if pos = new_pos - 1 is not visited
			if (distance[new_pos - 1] == 0) {
				// update : new_pos - 1
				distance[new_pos - 1] = new_count + 1;

				// add : node of new_pos - 1, new_count + 1
				enque(new_pos - 1, new_count + 1);
			}

			// if pos = new_pos + 1 is not visited
			if (distance[new_pos + 1] == 0) {
				// update : new_pos + 1
				distance[new_pos + 1] = new_count + 1;

				// add : node of new_pos + 1, new_count + 1
				enque(new_pos + 1, new_count + 1);
			}

			// if pos = new_pos * 2 is not visited
			if (distance[new_pos * 2] == 0) {
				// update : new_pos * 2
				distance[new_pos * 2] = new_count + 1;

				// add : node of new_pos * 2, new_count + 1
				enque(new_pos * 2, new_count + 1);
			}
		}

		printf("%d", distance[K]);
	}

	return 0;
}

void enque(int cur_pos, int count) {
	// update : rear
	rear = (rear + 1) % MAX;

	// create : new node
	node temp;
	temp.cur_pos = cur_pos;
	temp.count = count;

	// enque
	queue[rear] = temp;
}

node deque() {
	// update : front
	front = (front + 1) % MAX;	

	// deque
	return queue[front];
}

// return min(a, b)
int get_min(int a, int b) {
	if (a > b) {
		return b;
	}
	else {
		return a;
	}
}